runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java

```diff
@@ -0,0 +1,151 @@
+      GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+      Supplier<String> idGenerator,
+      ControlClientPool.Source clientSource) {
+    this.docker = docker;
+    this.controlServiceServer = controlServiceServer;
+    this.loggingServiceServer = loggingServiceServer;
+    this.retrievalServiceServer = retrievalServiceServer;
+    this.provisioningServiceServer = provisioningServiceServer;
+    this.idGenerator = idGenerator;
+    this.clientSource = clientSource;
+  }
+
+  /** Creates a new, active {@link RemoteEnvironment} backed by a local Docker container. */
+  @Override
+  public RemoteEnvironment createEnvironment(Environment environment) throws Exception {
+    String workerId = idGenerator.get();
+
+    // Prepare docker invocation.
+    Path workerPersistentDirectory = Files.createTempDirectory("worker_persistent_directory");
+    Path semiPersistentDirectory = Files.createTempDirectory("semi_persistent_dir");
+    String containerImage = environment.getUrl();
+    // TODO: https://issues.apache.org/jira/browse/BEAM-4148 The default service address will not
+    // work for Docker for Mac.
+    String loggingEndpoint = loggingServiceServer.getApiServiceDescriptor().getUrl();
+    String artifactEndpoint = retrievalServiceServer.getApiServiceDescriptor().getUrl();
+    String provisionEndpoint = provisioningServiceServer.getApiServiceDescriptor().getUrl();
+    String controlEndpoint = controlServiceServer.getApiServiceDescriptor().getUrl();
+    List<String> args =
+        Arrays.asList(
+            "-v",
+            // TODO: Mac only allows temporary mounts under /tmp by default (as of 17.12).
+            String.format("%s:%s", workerPersistentDirectory, semiPersistentDirectory),
+            // NOTE: Host networking does not work on Mac, but the command line flag is accepted.
+            "--network=host",
+            containerImage,
+            String.format("--id=%s", workerId),
+            String.format("--logging_endpoint=%s", loggingEndpoint),
+            String.format("--artifact_endpoint=%s", artifactEndpoint),
+            String.format("--provision_endpoint=%s", provisionEndpoint),
+            String.format("--control_endpoint=%s", controlEndpoint),
+            String.format("--semi_persist_dir=%s", semiPersistentDirectory));
+
+    // Wrap the blocking call to clientSource.get in case an exception is thrown.
+    String containerId = null;
+    InstructionRequestHandler instructionHandler = null;
+    try {
+      containerId = docker.runImage(containerImage, args);
+      // Wait on a client from the gRPC server.
+      while (instructionHandler == null) {
+        try {
+          instructionHandler = clientSource.get(workerId, Duration.ofMinutes(2));
+        } catch (TimeoutException timeoutEx) {
+          LOG.info("Waiting for docker environment: {}", environment.getUrl());
```

Comment:
<!--new_thread; commit:17af74e5551a5bf902b5497063e86d1f131128d6; resolved:0-->
Elaborate the log? E.g. "Still waiting for startup of environment {environment.getUrl()} in container {workerId}"